SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp

```diff
@@ -443,62 +435,54 @@ void UGlobalStateManager::SetAcceptingPlayers(bool bInAcceptingPlayers)
 	NetDriver->Connection->SendComponentUpdate(GlobalStateManagerEntityId, &Update);
 }
 
-void UGlobalStateManager::SetCanBeginPlay(bool bInCanBeginPlay)
+void UGlobalStateManager::SetAuthBeginPlayCalled(const bool bInAuthBeginPlayCalled)
 {
-	if (!NetDriver->StaticComponentView->HasAuthority(GlobalStateManagerEntityId, SpatialConstants::STARTUP_ACTOR_MANAGER_COMPONENT_ID))
-	{
-		UE_LOG(LogGlobalStateManager, Warning, TEXT("Tried to set CanBeginPlay on the GSM but this worker does not have authority."));
-		return;
-	}
+	check(NetDriver->StaticComponentView->HasAuthority(GlobalStateManagerEntityId, SpatialConstants::STARTUP_ACTOR_MANAGER_COMPONENT_ID));
 
 	Worker_ComponentUpdate Update = {};
 	Update.component_id = SpatialConstants::STARTUP_ACTOR_MANAGER_COMPONENT_ID;
 	Update.schema_type = Schema_CreateComponentUpdate(SpatialConstants::STARTUP_ACTOR_MANAGER_COMPONENT_ID);
 	Schema_Object* UpdateObject = Schema_GetComponentUpdateFields(Update.schema_type);
 
-	// Set CanBeginPlay on GSM
-	Schema_AddBool(UpdateObject, SpatialConstants::STARTUP_ACTOR_MANAGER_CAN_BEGIN_PLAY_ID, static_cast<uint8_t>(bInCanBeginPlay));
+	Schema_AddBool(UpdateObject, SpatialConstants::STARTUP_ACTOR_MANAGER_AUTH_BEGIN_PLAY_CALLED_ID, static_cast<uint8_t>(bInAuthBeginPlayCalled));
 
-	bCanBeginPlay = bInCanBeginPlay;
+	bAuthBeginPlayCalled = bInAuthBeginPlayCalled;
 	NetDriver->Connection->SendComponentUpdate(GlobalStateManagerEntityId, &Update);
 }
 
-void UGlobalStateManager::TryTriggerBeginPlay()
+void UGlobalStateManager::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
 {
-	if (bCanBeginPlay && NetDriver->EntityPool != nullptr && NetDriver->EntityPool->IsReady())
-	{
-		TriggerBeginPlay();
+	const bool bWorkerAuthority = AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE;
+	UE_LOG(LogGlobalStateManager, Log, TEXT("Authority over the GSM component %d has changed. This worker %s authority."), AuthOp.component_id, bWorkerAuthority ? TEXT("now has") : TEXT ("does not have"));
 
-		// Start accepting players only AFTER we've triggered BeginPlay
-		if (NetDriver->StaticComponentView->HasAuthority(GlobalStateManagerEntityId, SpatialConstants::DEPLOYMENT_MAP_COMPONENT_ID))
+	if (bWorkerAuthority)
+	{
+		switch (AuthOp.component_id)
 		{
-			SetAcceptingPlayers(true);
+			case SpatialConstants::DEPLOYMENT_MAP_COMPONENT_ID:
+				GlobalStateManagerEntityId = AuthOp.entity_id;
+				SetAcceptingPlayers(true);
+				break;
+			case SpatialConstants::SINGLETON_MANAGER_COMPONENT_ID:
+				ExecuteInitialSingletonActorReplication();
+				break;
+			default:
+				break;
 		}
 	}
 }
 
-void UGlobalStateManager::AuthorityChanged(bool bWorkerAuthority, Worker_EntityId CurrentEntityID)
+bool UGlobalStateManager::HandlesComponent(const Worker_ComponentId ComponentId) const
```

Comment:
I think this is a great function to have but the name could potentially be more descriptive but I don't have a better suggestion at the moment so I think it's fine to keep it as is.

Reply:
IsResponsibleForComponent?  Not sure if that's much better